src/stream/src/executor/mview/materialize.rs

```diff
@@ -558,189 +558,161 @@ pub struct MaterializeCache<SD> {
 
-                    if update_cache {
-                        match conflict_behavior {
-                            ConflictBehavior::Overwrite
-                            | ConflictBehavior::IgnoreConflict
-                            | ConflictBehavior::DoUpdateIfNotNull => {
-                                self.data.push(key, Some(CompactedRow { row: value }));
+                            if need_overwrite {
+                                let mut row_deserialized_vec =
+                                    old_row_deserialized.into_inner().into_vec();
+                                replace_if_not_null(
+                                    &mut row_deserialized_vec,
+                                    new_row_deserialized,
+                                );
+                                let updated_row = OwnedRow::new(row_deserialized_vec);
+                                let updated_row_bytes =
+                                    Bytes::from(row_serde.serializer.serialize(updated_row));
+
+                                change_buffer.update(
+                                    key.clone(),
+                                    old_row.row.clone(),
+                                    updated_row_bytes.clone(),
+                                );
+                                self.lru_cache.put(
+                                    key.clone(),
+                                    Some(CompactedRow {
+                                        row: updated_row_bytes,
+                                    }),
+                                );
                             }
-
-                            _ => unreachable!(),
                         }
-                    }
+                        _ => unreachable!(),
+                    };
                 }
 
                 Op::Delete | Op::UpdateDelete => {
                     match conflict_behavior {
-                        ConflictBehavior::Overwrite
-                        | ConflictBehavior::IgnoreConflict
-                        | ConflictBehavior::DoUpdateIfNotNull => {
-                            // delete a nonexistent value
-                            if let Some(old_row) = self.force_get(&key) {
-                                fixed_changes().delete(key.clone(), old_row.row.clone());
+                        checked_conflict_behaviors!() => {
+                            if let Some(old_row) = self.get_expected(&key) {
+                                change_buffer.delete(key.clone(), old_row.row.clone());
+                                // put a None into the cache to represent deletion
+                                self.lru_cache.put(key, None);
+                            } else {
+                                // delete a non-existent value
+                                // this is allowed in the case of mview conflict, so ignore
                             };
                         }
                         _ => unreachable!(),
```

Comment:
Is this actually reachable if there's inconsistency?

Reply:
No. The whole `MaterializeCache` is only used when `conflict_behavior` matches `checked_conflict_behaviors!()`.